File: go/internal/agent/reviewer/standards_test.go
package reviewer

import (
	"context"
	"strings"
	"testing"

	"automation-agent/internal/githubapi"
)

func TestMatchStandards(t *testing.T) {
	entries := []githubapi.TreeEntry{
		{Path: "AGENTS.md", SHA: "a", Type: "blob"},
		{Path: "internal/AGENTS.md", SHA: "b", Type: "blob"},
		{Path: ".cursor/rules/go.mdc", SHA: "c", Type: "blob"},
		{Path: ".cursor/rules", SHA: "d", Type: "tree"}, // a directory, not a doc
		{Path: "main.go", SHA: "e", Type: "blob"},       // not a standards file
	}
	globs := []string{"AGENTS.md", "**/AGENTS.md", ".cursor/rules/**"}
	got := matchStandards(entries, globs)
	want := map[string]bool{"AGENTS.md": true, "internal/AGENTS.md": true, ".cursor/rules/go.mdc": true}
	if len(got) != len(want) {
		t.Fatalf("matched %d, want %d: %+v", len(got), len(want), got)
	}
	for _, m := range got {
		if !want[m.Path] {
			t.Errorf("unexpected match %q", m.Path)
		}
		if m.Type != "blob" {
			t.Errorf("matched a non-blob: %+v", m)
		}
	}
}

func TestStandardsCacheKeyChangesWithSHA(t *testing.T) {
	base := []githubapi.TreeEntry{{Path: "AGENTS.md", SHA: "v1"}}
	changed := []githubapi.TreeEntry{{Path: "AGENTS.md", SHA: "v2"}}
	if standardsCacheKey("o", "r", base) == standardsCacheKey("o", "r", changed) {
		t.Error("a changed blob SHA must change the cache key (else stale rules)")
	}
	if standardsCacheKey("o", "r", base) != standardsCacheKey("o", "r", base) {
		t.Error("same inputs must produce the same key")
	}
}

func TestParseRules(t *testing.T) {
	t.Run("prose-wrapped, dedup, drops blank", func(t *testing.T) {
		raw := "Here are the rules:\n```json\n[" +
			`{"id":"R1","dimension":"security","summary":"validate input","source":"SECURITY.md"},` +
			`{"id":"R1","dimension":"x","summary":"dup id dropped","source":"x"},` +
			`{"id":"R2","summary":"  ","source":"y"},` + // blank summary dropped
			`{"id":"R3","dimension":"vibes","summary":"prefer composition","source":"AGENTS.md"}` +
			"]\n```"
		got := parseRules(raw)
		if len(got) != 2 {
			t.Fatalf("got %d rules, want 2 (dup + blank dropped): %+v", len(got), got)
		}
		if got[0].ID != "R1" || got[0].Dimension != DimSecurity {
			t.Errorf("rule 0 = %+v", got[0])
		}
		if got[1].ID != "R3" || got[1].Dimension != DimOther { // "vibes" normalizes to other
			t.Errorf("rule 1 = %+v", got[1])
		}
	})
	t.Run("malformed yields nil", func(t *testing.T) {
		for _, raw := range []string{"", "no json", "[{broken", "{\"not\":\"array\"}", "[]"} {
			if got := parseRules(raw); got != nil {
				t.Errorf("parseRules(%q) = %+v, want nil", raw, got)
			}
		}
	})
}

func TestStandardsMenuAndLookup(t *testing.T) {
	std := buildStandards(
		[]Rule{{ID: "R1", Dimension: DimSecurity, Summary: "validate input", Source: "SECURITY.md"}},
		map[string]string{"SECURITY.md": "full security doc text"},
		[]string{"SECURITY.md"},
	)
	if std.empty() {
		t.Fatal("std must not be empty")
	}
	if !std.validID("R1") || std.validID("R9") {
		t.Error("validID wrong")
	}
	if std.ruleDoc("R1") != "full security doc text" {
		t.Errorf("ruleDoc(R1) = %q", std.ruleDoc("R1"))
	}
	if std.ruleDoc("R9") != "" {
		t.Error("unknown id must yield empty doc")
	}
	for _, want := range []string{"R1", "security", "validate input", "SECURITY.md"} {
		if !strings.Contains(std.menu(), want) {
			t.Errorf("menu missing %q:\n%s", want, std.menu())
		}
	}
}

func TestGateCitations(t *testing.T) {
	std := buildStandards(
		[]Rule{{ID: "R1", Dimension: DimPatternViolation, Summary: "s", Source: "AGENTS.md"}},
		map[string]string{"AGENTS.md": "doc"}, []string{"AGENTS.md"},
	)
	findings := []Finding{
		{Dimension: DimPatternViolation, Severity: SeverityMajor, Message: "cited", RuleID: "R1"}, // kept
		{Dimension: DimPatternViolation, Severity: SeverityMajor, Message: "uncited"},             // gated
		{Dimension: DimArchitecture, Severity: SeverityMajor, Message: "bad id", RuleID: "R9"},    // gated (unknown id)
		{Dimension: DimSecurity, Severity: SeverityCritical, Message: "sqli"},                     // never gated
	}

	t.Run("nitpick mode demotes uncited conformance", func(t *testing.T) {
		e := NewEngine(Deps{StandardsEnabled: true}) // UncitedDrop false -> nitpick
		got := e.gateCitations(append([]Finding(nil), findings...), std)
		if len(got) != 4 {
			t.Fatalf("nitpick mode keeps all, got %d", len(got))
		}
		if got[0].Severity != SeverityMajor { // cited conformance untouched
			t.Errorf("cited finding altered: %+v", got[0])
		}
		if got[1].Severity != SeverityNitpick || got[2].Severity != SeverityNitpick {
			t.Errorf("uncited conformance not demoted: %+v %+v", got[1], got[2])
		}
		if got[3].Severity != SeverityCritical { // security never gated
			t.Errorf("non-conformance finding altered: %+v", got[3])
		}
	})

	t.Run("drop mode removes uncited conformance", func(t *testing.T) {
		e := NewEngine(Deps{StandardsEnabled: true, UncitedDrop: true})
		got := e.gateCitations(append([]Finding(nil), findings...), std)
		if len(got) != 2 { // the cited conformance + the security finding
			t.Fatalf("drop mode = %d findings, want 2", len(got))
		}
	})

	t.Run("disabled passes everything through", func(t *testing.T) {
		e := NewEngine(Deps{StandardsEnabled: false, UncitedDrop: true})
		got := e.gateCitations(append([]Finding(nil), findings...), std)
		if len(got) != 4 {
			t.Errorf("standards off must not gate, got %d", len(got))
		}
	})
}

func TestStandardsToolsPresence(t *testing.T) {
	std := buildStandards([]Rule{{ID: "R1", Summary: "s"}}, map[string]string{}, []string{"AGENTS.md"})
	tools, err := standardsTools(std)
	if err != nil || len(tools) != 1 {
		t.Fatalf("standardsTools(non-empty) = %d tools, err %v; want 1", len(tools), err)
	}
	empty, err := standardsTools(nil)
	if err != nil || empty != nil {
		t.Errorf("standardsTools(nil) = %v, err %v; want nil/nil", empty, err)
	}
}

func TestDiscoverStandards(t *testing.T) {
	rulesJSON := `[{"id":"R1","dimension":"pattern_violation","summary":"wrap errors","source":"AGENTS.md"}]`
	newEngine := func(gh *fakeGH) *Engine {
		llm := fakeLLM{json: rulesJSON}
		return NewEngine(Deps{
			Enabled: true, GH: gh, BaseLLM: llm, CodeLLM: llm,
			StandardsEnabled: true, StandardsGlobs: []string{"AGENTS.md"}, StandardsMaxBytes: 1 << 20,
		})
	}

	t.Run("discovers, distills, caches", func(t *testing.T) {
		gh := &fakeGH{
			tree:     []githubapi.TreeEntry{{Path: "AGENTS.md", SHA: "s1", Type: "blob"}},
			contents: map[string]string{"AGENTS.md": "wrap errors with %w"},
		}
		e := newEngine(gh)
		std := e.discoverStandards(context.Background(), "o", "r", "head", nil)
		if std.empty() || len(std.rules) != 1 || std.rules[0].ID != "R1" {
			t.Fatalf("discovered std = %+v", std)
		}
		if len(std.sourceList()) != 1 || std.sourceList()[0] != "AGENTS.md" {
			t.Errorf("sources = %v", std.sourceList())
		}
		// Second call (same tree SHAs) is a cache hit: same pointer, no re-distill.
		if again := e.discoverStandards(context.Background(), "o", "r", "head", nil); again != std {
			t.Error("second discovery must hit the cache (same *standards)")
		}
	})

	t.Run("disabled yields nil", func(t *testing.T) {
		e := NewEngine(Deps{Enabled: true, GH: &fakeGH{}, StandardsEnabled: false})
		if std := e.discoverStandards(context.Background(), "o", "r", "head", nil); std != nil {
			t.Errorf("disabled discovery = %+v, want nil", std)
		}
	})

	t.Run("no matching docs yields nil", func(t *testing.T) {
		gh := &fakeGH{tree: []githubapi.TreeEntry{{Path: "main.go", SHA: "s", Type: "blob"}}}
		if std := newEngine(gh).discoverStandards(context.Background(), "o", "r", "head", nil); std != nil {
			t.Errorf("no standards docs = %+v, want nil", std)
		}
	})

	t.Run("tree error is best-effort nil", func(t *testing.T) {
		gh := &fakeGH{treeErr: context.DeadlineExceeded}
		if std := newEngine(gh).discoverStandards(context.Background(), "o", "r", "head", nil); std != nil {
			t.Errorf("tree error must degrade to nil, got %+v", std)
		}
	})

	// A truncated tree means discovery may have missed convention files; gating against the partial
	// set is worse than a generic review, so degrade to nil (and don't cache).
	t.Run("truncated tree degrades to generic nil", func(t *testing.T) {
		gh := &fakeGH{
			tree:      []githubapi.TreeEntry{{Path: "AGENTS.md", SHA: "s1", Type: "blob"}},
			truncated: true,
			contents:  map[string]string{"AGENTS.md": "wrap errors with %w"},
		}
		if std := newEngine(gh).discoverStandards(context.Background(), "o", "r", "head", nil); std != nil {
			t.Errorf("truncated tree must degrade to nil, got %+v", std)
		}
	})

	// A fetch failure leaves the rule set incomplete; degrade to generic for this round and do not
	// cache, so a later event retries the full set rather than serving a memoized partial.
	t.Run("partial fetch failure degrades to nil, uncached", func(t *testing.T) {
		gh := &fakeGH{
			tree: []githubapi.TreeEntry{
				{Path: "AGENTS.md", SHA: "s1", Type: "blob"},
				{Path: "CONTRIBUTING.md", SHA: "s2", Type: "blob"},
			},
			contents: map[string]string{"AGENTS.md": "wrap errors"}, // CONTRIBUTING.md fetch fails
		}
		e := NewEngine(Deps{
			Enabled: true, GH: gh, BaseLLM: fakeLLM{json: rulesJSON}, CodeLLM: fakeLLM{json: rulesJSON},
			StandardsEnabled: true, StandardsGlobs: []string{"AGENTS.md", "CONTRIBUTING.md"}, StandardsMaxBytes: 1 << 20,
		})
		if std := e.discoverStandards(context.Background(), "o", "r", "head", nil); std != nil {
			t.Errorf("partial fetch must degrade to nil, got %+v", std)
		}
		// Not memoized: once the missing doc resolves, the next event must build the full set.
		gh.contents["CONTRIBUTING.md"] = "prefer composition"
		if std := e.discoverStandards(context.Background(), "o", "r", "head", nil); std.empty() {
			t.Error("after the fetch resolves, discovery must build (partial was not cached)")
		}
	})

	// A per-directory instruction file applies only to its own touched module; root files always
	// apply; an untouched module's file is excluded.
	t.Run("per-module instruction file scoped to touched dirs", func(t *testing.T) {
		gh := &fakeGH{
			tree: []githubapi.TreeEntry{
				{Path: "AGENTS.md", SHA: "s0", Type: "blob"},
				{Path: "internal/foo/AGENTS.md", SHA: "s1", Type: "blob"},
				{Path: "internal/bar/AGENTS.md", SHA: "s2", Type: "blob"},
			},
			contents: map[string]string{
				"AGENTS.md":              "root",
				"internal/foo/AGENTS.md": "foo",
				"internal/bar/AGENTS.md": "bar",
			},
		}
		e := NewEngine(Deps{
			Enabled: true, GH: gh, BaseLLM: fakeLLM{json: rulesJSON}, CodeLLM: fakeLLM{json: rulesJSON},
			StandardsEnabled: true, StandardsGlobs: []string{"AGENTS.md", "**/AGENTS.md"}, StandardsMaxBytes: 1 << 20,
		})
		std := e.discoverStandards(context.Background(), "o", "r", "head", []githubapi.PRFile{{Path: "internal/foo/x.go"}})
		if std.empty() {
			t.Fatal("expected standards")
		}
		got := std.sourceList()
		want := map[string]bool{"AGENTS.md": true, "internal/foo/AGENTS.md": true}
		if len(got) != 2 || !want[got[0]] || !want[got[1]] {
			t.Errorf("scoped sources = %v, want root + internal/foo only (internal/bar excluded)", got)
		}
	})
}

File: kotlin/src/test/kotlin/com/automation/agent/agent/reviewer/StandardsTest.kt
package com.automation.agent.agent.reviewer

import com.automation.agent.agent.setup.assistantText
import com.automation.agent.githubapi.CheckResult
import com.automation.agent.githubapi.CheckRunInput
import com.automation.agent.githubapi.PRFile
import com.automation.agent.githubapi.ReviewCommentRef
import com.automation.agent.githubapi.ReviewInput
import com.automation.agent.githubapi.Tree
import com.automation.agent.githubapi.TreeEntry
import com.google.adk.kt.models.LlmRequest
import com.google.adk.kt.models.LlmResponse
import com.google.adk.kt.models.Model
import com.google.adk.kt.types.Content
import com.google.adk.kt.types.FunctionCall
import com.google.adk.kt.types.Part
import com.google.adk.kt.types.Role
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flowOf

private class StdModel(private val json: String) : Model {
    override val name: String = "fake"
    override fun generateContent(request: LlmRequest, stream: Boolean): Flow<LlmResponse> =
        flowOf(LlmResponse(content = assistantText(json)))
}

/** A fake GitHub client serving a canned tree + file contents for standards discovery. */
private class StdGh(
    private val tree: Tree,
    private val contents: Map<String, String> = emptyMap(),
) : GitHubClient {
    override suspend fun listPRFiles(owner: String, repo: String, num: Int) = emptyList<PRFile>()
    override suspend fun createReview(owner: String, repo: String, num: Int, input: ReviewInput) {}
    override suspend fun upsertMarkerComment(owner: String, repo: String, num: Int, marker: String, body: String) {}
    override suspend fun createCheckRun(owner: String, repo: String, input: CheckRunInput) {}
    override suspend fun listReviewComments(owner: String, repo: String, num: Int) = emptyList<ReviewCommentRef>()
    override suspend fun minimizeComment(subjectId: String) {}
    override suspend fun agentCheck(owner: String, repo: String, ref: String, checkName: String) = CheckResult(found = false)
    override suspend fun pullRequestHeadSha(owner: String, repo: String, num: Int) = ""
    override suspend fun tree(owner: String, repo: String, ref: String) = tree
    override suspend fun getFileContent(owner: String, repo: String, path: String, ref: String) = contents[path] ?: ""
}

private val RULE_JSON =
    "[{\"id\":\"R1\",\"dimension\":\"pattern_violation\",\"summary\":\"wrap errors\",\"source\":\"AGENTS.md\"}]"

/** A model that calls get_rule on its first turn, then returns [json] once the tool result is fed back. */
private class ToolThenTextModel(private val json: String) : Model {
    override val name: String = "fake"
    private var calls = 0
    override fun generateContent(request: LlmRequest, stream: Boolean): Flow<LlmResponse> {
        calls++
        return if (calls == 1) {
            flowOf(LlmResponse(content = Content(role = Role.MODEL, parts = listOf(Part(functionCall = FunctionCall(name = "get_rule", args = mapOf("id" to "R1"), id = "get_rule"))))))
        } else {
            flowOf(LlmResponse(content = assistantText(json)))
        }
    }
}

private fun stdEngine(
    gh: GitHubClient? = null,
    model: Model? = StdModel(RULE_JSON),
    enabled: Boolean = true,
    uncitedDrop: Boolean = false,
) = Engine(
    Deps(
        enabled = true, gh = gh, baseLlm = model, codeLlm = model,
        standardsEnabled = enabled, standardsGlobs = listOf("AGENTS.md", ".agents/standards/**"),
        standardsMaxBytes = 1_000_000, uncitedDrop = uncitedDrop,
    ),
)

private fun blob(path: String, sha: String = "s-$path") = TreeEntry(path = path, sha = sha, type = "blob")

class StandardsTest : BehaviorSpec({
    Given("glob matching over a tree") {
        Then("matchStandards keeps only matching blobs, sorted") {
            val entries = listOf(blob("src/Main.kt"), blob("AGENTS.md"), blob(".agents/standards/style.md"), TreeEntry("dir", "x", "tree"))
            matchStandards(entries, listOf("AGENTS.md", ".agents/standards/**")).map { it.path } shouldBe
                listOf(".agents/standards/style.md", "AGENTS.md")
        }
        Then("matchesGlob honors basename vs path globs") {
            val pats = compileStandardsGlobs(listOf("AGENTS.md", "vendor/**", ""))
            matchesGlob(pats, "kotlin/AGENTS.md") shouldBe true
            matchesGlob(pats, "vendor/x/y") shouldBe true
            matchesGlob(pats, "src/Main.kt") shouldBe false
        }
    }

    Given("per-module scoping") {
        Then("moduleScoped is true only for nested instruction files") {
            moduleScoped("AGENTS.md") shouldBe false
            moduleScoped("kotlin/AGENTS.md") shouldBe true
            moduleScoped("kotlin/src/config.yaml") shouldBe false
        }
        Then("touchedDirs walks every ancestor to the root") {
            touchedDirs(listOf(PRFile(path = "kotlin/src/Main.kt"))) shouldBe setOf("kotlin/src", "kotlin", ".")
        }
        Then("scopeToTouched drops instruction files for untouched modules, keeps root ones") {
            val matched = listOf(blob("AGENTS.md"), blob("kotlin/AGENTS.md"), blob("python/AGENTS.md"))
            val changed = listOf(PRFile(path = "kotlin/src/Main.kt"))
            scopeToTouched(matched, changed).map { it.path } shouldBe listOf("AGENTS.md", "kotlin/AGENTS.md")
        }
    }

    Given("the cache key") {
        Then("it changes when a matched blob SHA changes") {
            val a = standardsCacheKey("o", "r", listOf(blob("AGENTS.md", "sha1")))
            val b = standardsCacheKey("o", "r", listOf(blob("AGENTS.md", "sha2")))
            (a == b) shouldBe false
            standardsCacheKey("o", "r", listOf(blob("AGENTS.md", "sha1"))) shouldBe a
        }
    }

    Given("parseRules") {
        Then("it recovers a rule array from fenced prose and dedups ids") {
            val rules = parseRules("```json\n$RULE_JSON\n```")
            rules.size shouldBe 1
            rules[0].id shouldBe "R1"
            rules[0].dimension shouldBe Dimension.PATTERN_VIOLATION
        }
        Then("a garbled or empty distillation degrades to no rules") {
            parseRules("not json") shouldBe emptyList()
            parseRules("[]") shouldBe emptyList()
            parseRules("[{\"id\":\"\",\"summary\":\"x\"}]") shouldBe emptyList()
        }
    }

    Given("Standards + the get_rule tool") {
        val std = buildStandards(
            listOf(Rule("R1", Dimension.PATTERN_VIOLATION, "wrap errors", "AGENTS.md")),
            mapOf("AGENTS.md" to "the full standards text"),
            listOf("AGENTS.md"),
        )
        Then("the menu, validId, ruleDoc, and sourceList behave") {
            isEmpty(std) shouldBe false
            val s = std.shouldNotBeNull()
            s.menu() shouldContain "- R1 [pattern_violation] wrap errors (source: AGENTS.md)"
            s.validId("R1") shouldBe true
            s.validId("R9") shouldBe false
            s.ruleDoc("R1") shouldBe "the full standards text"
            s.ruleDoc("R9") shouldBe ""
            s.sourceList() shouldBe listOf("AGENTS.md")
        }
        Then("standardsTools serves the full text and self-wraps a lookup") {
            val tool = standardsTools(std).single()
            tool.declaration.name shouldBe "get_rule"
            tool.execute(mapOf("id" to "R1")) shouldBe mapOf("rule" to "the full standards text")
            tool.execute(mapOf("id" to "R9")) shouldBe mapOf("rule" to "")
            standardsTools(null) shouldBe emptyList()
        }
        Then("writeStandardsMenu injects the menu only when standards are present") {
            val withStd = StringBuilder("P").also { writeStandardsMenu(it, std) }.toString()
            withStd shouldContain "Repo standards (cite rule_id for conformance findings)"
            StringBuilder("P").also { writeStandardsMenu(it, null) }.toString() shouldBe "P"
        }
        Then("buildDistillerInstruction fences each doc under a Document label") {
            val instr = buildDistillerInstruction("PROMPT", mapOf("AGENTS.md" to "body ``` more"), listOf("AGENTS.md"))
            instr shouldContain "### Document: AGENTS.md"
            instr shouldContain "````" // fence sized past the embedded triple-backtick run
        }
    }

    Given("gateCitations") {
        val std = buildStandards(listOf(Rule("R1", Dimension.PATTERN_VIOLATION, "s", "AGENTS.md")), mapOf("AGENTS.md" to "t"), listOf("AGENTS.md"))
        val uncited = Finding(dimension = Dimension.PATTERN_VIOLATION, severity = Severity.MAJOR, message = "unanchored")
        val cited = Finding(dimension = Dimension.PATTERN_VIOLATION, severity = Severity.MAJOR, message = "anchored", ruleId = "R1")
        val security = Finding(dimension = Dimension.SECURITY, severity = Severity.MAJOR, message = "sec")
        Then("uncited-drop removes an unanchored conformance finding, keeps cited + non-conformance") {
            val kept = gateCitations(stdEngine(uncitedDrop = true), listOf(uncited, cited, security), std)
            kept.map { it.message } shouldBe listOf("anchored", "sec")
        }
        Then("uncited-nitpick demotes rather than drops") {
            val kept = gateCitations(stdEngine(uncitedDrop = false), listOf(uncited), std)
            kept.single().severity shouldBe Severity.NITPICK
        }
        Then("standards-off passes findings through untouched") {
            gateCitations(stdEngine(enabled = false), listOf(uncited), std) shouldBe listOf(uncited)
        }
    }

    Given("the standards cache") {
        Then("a miss then a stored (even null) value hits") {
            val c = StandardsCache()
            c.get("k").ok shouldBe false
            c.put("k", null)
            c.get("k").let { it.ok shouldBe true; it.std shouldBe null }
        }
    }

    Given("discoverStandards end-to-end") {
        val gh = StdGh(Tree(listOf(blob("AGENTS.md")), truncated = false), mapOf("AGENTS.md" to "docs"))
        Then("it discovers, distills, caches, and returns the rule set") {
            val e = stdEngine(gh = gh)
            val std = discoverStandards(e, "acme", "api", "sha", listOf(PRFile(path = "src/Main.kt"))).shouldNotBeNull()
            std.rules.single().id shouldBe "R1"
            // Second call is served from the cache (same key).
            discoverStandards(e, "acme", "api", "sha", listOf(PRFile(path = "src/Main.kt"))).shouldNotBeNull().rules.size shouldBe 1
        }
        Then("it degrades to generic when disabled, truncated, or nothing matches") {
            discoverStandards(stdEngine(gh = gh, enabled = false), "a", "b", "s", emptyList()) shouldBe null
            val trunc = StdGh(Tree(listOf(blob("AGENTS.md")), truncated = true))
            discoverStandards(stdEngine(gh = trunc), "a", "b", "s", emptyList()) shouldBe null
            val none = StdGh(Tree(listOf(blob("src/Main.kt")), truncated = false))
            discoverStandards(stdEngine(gh = none), "a", "b", "s", emptyList()) shouldBe null
        }
        Then("a distillation that yields no rules degrades to generic (uncached null retained)") {
            val e = stdEngine(gh = gh, model = StdModel("[]"))
            discoverStandards(e, "acme", "api", "sha", listOf(PRFile(path = "src/Main.kt"))) shouldBe null
        }
    }
})
